packages/api/src/incentives/milestones/milestoneService.ts

```diff
@@ -0,0 +1,459 @@
+      ) {
+        const groupedMilestones: GroupedMilestones = {
+          tvl: {
+            current: 0,
+            milestones: [],
+          },
+          transactions: {
+            current: 0,
+            milestones: [],
+          },
+          dex_volume: {
+            current: 0,
+            milestones: [],
+          },
+          wallet_downloads: {
+            current: 0,
+            milestones: [],
+          },
+        };
+
+        for (const milestone of milestonesToUse) {
+          const currentValue = parseFloat(milestone.currentValue);
+          const target = parseFloat(milestone.goal);
+          const rewardXrd = parseFloat(milestone.rewardXrd);
+          const milestoneType =
+            milestone.type as keyof typeof groupedMilestones;
+
+          if (groupedMilestones[milestoneType]) {
+            // Update current value to the highest among milestones of this type
+            groupedMilestones[milestoneType].current = Math.max(
+              groupedMilestones[milestoneType].current,
+              currentValue,
+            );
+
+            groupedMilestones[milestoneType].milestones.push({
+              target,
+              reward: rewardXrd,
+              achieved: milestone.isAchieved,
+            });
+          }
+        }
+
+        // Sort milestones by target value
+        for (const type of Object.keys(groupedMilestones) as Array<
+          keyof typeof groupedMilestones
+        >) {
+          groupedMilestones[type].milestones.sort(
+            (a, b) => a.target - b.target,
+          );
+        }
+
+        return groupedMilestones;
+      });
+
+      return {
+        /**
+         * Create a new milestone
+         */
+        createMilestone: Effect.fn(function* (input: CreateMilestoneInput) {
+          const result = yield* Effect.tryPromise({
+            try: () =>
+              db
+                .insert(milestones)
+                .values({
+                  seasonId: input.seasonId,
+                  type: input.type,
+                  goal: input.goal,
+                  rewardXrd: input.rewardXrd,
+                })
+                .returning(),
+            catch: (error) => new DbError(error),
+          });
+
+          return result[0];
+        }),
+
+        /**
+         * Get milestone progress data formatted for frontend components
+         */
+        getMilestoneProgress: Effect.fn(function* (seasonId: string) {
+          // Get current milestones
+          const allMilestones = yield* Effect.tryPromise({
+            try: () =>
+              db
+                .select()
+                .from(milestones)
+                .where(eq(milestones.seasonId, seasonId))
+                .orderBy(milestones.type, milestones.goal),
+            catch: (error) => new DbError(error),
+          });
+
+          // Group milestones by type for staleness checking
+          const tvlMilestones = allMilestones.filter((m) => m.type === 'tvl');
```

Comment:
use `groupBy` instead of running multiple filter operations

```
import { groupBy } from 'effect/Array';

const { 
  tvl: tvlMilestones, 
  transactions: transactionMilestones, 
  dex_volume: dexVolumeMilestones 
} = groupBy(allMilestones, (m) => m.type);
```